File: 5-Problem-A-Day/95. Unique Binary Search Trees II.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    private LinkedList<TreeNode> generate(int start, int end) {
        LinkedList<TreeNode> allTrees = new LinkedList<TreeNode>();
        if (start > end) {
            allTrees.add(null);
            return allTrees;
        }
        for (int i = start; i <= end; i++) {
            LinkedList<TreeNode> leftTrees = generate(start, i - 1);
            LinkedList<TreeNode> rightTrees = generate(i + 1, end);
            for (TreeNode l : leftTrees) {
                for (TreeNode r : rightTrees) {
                    TreeNode curr = new TreeNode(i);
                    curr.left = l;
                    curr.right = r;
                    allTrees.add(curr);
                }
            }
        }
        return allTrees;
    }

    public List<TreeNode> generateTrees(int n) {
        if (n == 0)
            return new LinkedList<TreeNode>();
        return generate(1, n);
    }
}

File: 2023/95. Unique Binary Search Trees II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateSubtrees(int start, int end) {
        if (start > end) {
            return {NULL};
        }
        vector<TreeNode*> ret;
        for (int i = start; i <= end; i++) {
            vector<TreeNode*> lefts = generateSubtrees(start, i - 1);
            vector<TreeNode*> rights = generateSubtrees(i + 1, end);
            for (TreeNode* left : lefts) {
                for (TreeNode* right : rights) {
                    TreeNode* root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    ret.push_back(root);
                }

            }
        }
        return ret;
    }

    vector<TreeNode*> generateTrees(int n) {
        return generateSubtrees(1, n);
    }
};